### PR TITLE
Taking a reference to an optional value now produces an optionally-typed reference

### LINKED ISSUE #747: BUG: Should not be able to statically access the fields of a optionally casted reference
### Problem

When creating a reference to an object stored in a dictionary using `as?`, the checker does not report a static error when trying to access a field of the optional reference.

### Steps to Reproduce

1. Copy this contract
```
pub contract TestCollection {
    pub resource IDRes {
        pub let id: UInt32
        init() {
            self.id = 0
        }
    }

    pub resource Collection {
        access(self) var idResources: @{String: IDRes}
        init() {
            self.idResources <- {}
        }

        destroy() {
            destroy self.idResources
        }

        pub fun borrowID(_ key: String, _ id: UInt32): UInt32 {
            let idref = &self.idResources[key] as? &IDRes

            let id = idref.id

            return id
        }
    }

    init() {
        let col <- create Collection()

        let del = col.borrowID("hello", 1)

        destroy col
    }
}
```
2. Deploy the contract and see this error:
```
WARN[0004] ERR [08e509] Execution failed:
error: dereference failed
  --> 01cf0e2f2f715450.TestCollection:23:21
   |
23 |             let id = idref.id
```

### Acceptance Criteria

It should cause a static error because it is trying to access the field of an optional, right?

### PR DESCRIPTION
**Pull request by dsainati1**
Closes #747

## Description

It was possible to obtain a reference to an optional by using a dictionary access index expression, which could then have its fields accessed as if it were not optional, leading to the bug in the linked issue. 

This removes the ability to create a reference to an optional this way. Instead, to create a reference to the contents of a dictionary, one can create an optional reference like so:

```cadence
let dict: {String : S} = {}
let s = &dict[""] as &S? // s has type &S?
```

When the expression on the left of the `as` is optional, the target type on the right side must also be optional to match.

This is described in more detail in this FLIP: https://github.com/onflow/flow/pull/722 
______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

**Comment by dsainati1** [comment]
Note that this breaks the Topshot contract test, which implies that the existing Topshot contract would rely on the current, buggy behavior. I am unsure what the best way to resolve this is, or whether it simply indicates that this approach is flawed.

**Comment by turbolent** [comment]
This was not a bug, but a feature. I still think it makes sense to remove it, but we should pitch it to the community and then propose a FLIP before merging this, as it is a breaking change and some contracts rely on this feature.

Could you please remove the `Bugfix` label and add the  `Breaking Change` label?

File: runtime/interpreter/interpreter_expression.go
Comment (turbolent) [inline runtime/interpreter/interpreter_expression.go:822]: I'm not quite sure, but I think the type check shouldn't be on the dynamic type of the value, but the static type of the referenced expression

File: runtime/sema/check_reference_expression.go
Comment (turbolent) [inline runtime/sema/check_reference_expression.go:42]: Maybe just mention this as an example and document the general case:

 ```suggestion
		// Reference expressions may reference a value which has an optional type.
		// For example, the result of indexing into a dictionary is an optional:
		//
		// let ints: {Int: String} = {0: "zero"}
		// let ref: &T? = &ints[0] as &T?   // read as (&T)?
		//
		// In this case the reference expression's type is an optional type.
		// Unwrap it one level to get the actual reference type
```

File: runtime/sema/check_reference_expression.go
Comment (turbolent) [inline runtime/sema/check_reference_expression.go:72]:
```suggestion
		// If the referenced type is optional, then the return type is optional
```

File: runtime/sema/check_reference_expression.go
Comment (turbolent) [inline runtime/sema/check_reference_expression.go:77]: This if-statement looks like dead code now: It is the same code as above (`referencedType.(*OptionalType)`), and cannot occur anymore, as `referencedType` is unwrapped and re-assigned above.

I think this statement can be removed, and the error type `OptionalTypeReferenceError` too.

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:1050]: Maybe assert the type of `ref` to be an optional of a reference to `Int`

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:1012]: Maybe also assert the type of `ref` to be an optional of a reference to `Int`

File: runtime/tests/interpreter/interpreter_test.go
Comment (turbolent) [inline runtime/tests/interpreter/interpreter_test.go:8037]: Please keep the spaces in string literals, i.e. revert the changes

**Review by turbolent — COMMENTED** [review]
Nice work! 👏

File: runtime/tests/interpreter/interpreter_test.go
Comment (SupunS) [inline runtime/tests/interpreter/interpreter_test.go:9514]:
```suggestion
          fun present(): &Int {
              let x: Int? = 1
              let y = &x as &Int?
              return y!
          }

          fun absent(): &Int {
              let x: Int? = nil
              let y = &x as &Int?
              return y!
          }
```

File: runtime/tests/checker/reference_test.go
Comment (SupunS) [inline runtime/tests/checker/reference_test.go:1165]: 👌

File: runtime/sema/check_reference_expression.go
Comment (SupunS) [inline runtime/sema/check_reference_expression.go:46]: maybe leave a blank line after the block
```suggestion
		}

```

**Review by SupunS — COMMENTED** [review]
Once the suggestions are addressed, PR looks good to me!

File: runtime/interpreter/interpreter_expression.go
Comment (dsainati1) [reply runtime/interpreter/interpreter_expression.go:822]: We do need to do this switch on the dynamic type so we know whether or not to move the `*SomeValue` constructor. However you are right in that this check should only occur if the static type of the expression is itself optional.

**Comment by dsainati1** [comment]
The FLIP proposing this change has been approved and merged.

File: runtime/missingmember_test.go
Comment (turbolent) [inline runtime/missingmember_test.go:848]: Even though this makes the code valid, and the code itself is just a test case for something else and was provided by the community, we should maybe still update this code so it gracefully handles the case and to avoid developers seeing this and copying it as-is.

Concretely, in the cases where the functions return an optional (e.g. here, but not in the function above), maybe instead use the nil-coalescing operator `??` to return `nil` if the dictionary lookup fails, instead of aborting the execution:

```suggestion
               let ref = &self.ownedNFTs[id] as auth &NonFungibleToken.NFT?
               return ref ?? ref as! &GarmentNFT.NFT
```

File: runtime/interpreter/interpreter_expression.go
Comment (turbolent) [inline runtime/interpreter/interpreter_expression.go:827]: 👍

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:1163]:
```suggestion
			pub let foo: Number
			init() {
				self.foo = 0
			}
		}
		let dict: {String: S} = {}
```

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:1184]:
```suggestion
			pub let foo: Number
			init() {
				self.foo = 0
			}
		}
		let dict: {String: S} = {}
```

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:1199]:
```suggestion
			pub let foo: Number
```

**Review by turbolent — APPROVED** [review]
Looks good, great work! 👏 Just some minor style suggestions, good to merge after

**Pull request merged**
